File: javaproject/app/src/main/java/com/marcosvbras/robomarket/flows/profile/viewmodel/EditProfileViewModel.java
package com.marcosvbras.robomarket.flows.profile.viewmodel;

import android.annotation.SuppressLint;
import android.databinding.ObservableField;
import android.text.TextUtils;
import android.util.Patterns;
import android.view.View;
import android.widget.AdapterView;

import com.marcosvbras.robomarket.R;
import com.marcosvbras.robomarket.app.App;
import com.marcosvbras.robomarket.business.domain.User;
import com.marcosvbras.robomarket.business.model.UserModel;
import com.marcosvbras.robomarket.interfaces.BaseActivityCallback;
import com.marcosvbras.robomarket.utils.Constants;
import com.marcosvbras.robomarket.utils.ErrorObservable;
import com.marcosvbras.robomarket.viewmodels.BaseViewModel;

import java.util.Random;

import io.reactivex.disposables.Disposable;

public class EditProfileViewModel extends BaseViewModel {

    private BaseActivityCallback activityCallback;
    private Disposable disposable;
    private User user;
    private UserModel userModel;
    public ObservableField<String> email = new ObservableField<>();
    public ObservableField<String> username = new ObservableField<>();
    public ObservableField<String> name = new ObservableField<>();
    public ObservableField<String> phone = new ObservableField<>();
    public ObservableField<String> address = new ObservableField<>();
    public ObservableField<String> password = new ObservableField<>();
    public ObservableField<String> newPassword = new ObservableField<>();
    public ObservableField<String> confirmPassword = new ObservableField<>();
    public ObservableField<String> genre = new ObservableField<>();
    public ObservableField<String> avatarUrl = new ObservableField<>();
    public ErrorObservable emailFieldError = new ErrorObservable();
    public ErrorObservable newPasswordFieldError = new ErrorObservable();
    public ErrorObservable confirmPasswordFieldError = new ErrorObservable();
    public ErrorObservable usernameFieldError = new ErrorObservable();
    public ErrorObservable nameFieldError = new ErrorObservable();

    public EditProfileViewModel(BaseActivityCallback activityCallback) {
        this.activityCallback = activityCallback;
        setUserData();
        userModel = new UserModel();
    }

    private void setUserData() {
        user = (User) App.getInstance().getUser().clone();
        email.set(user.getEmail());
        username.set(user.getUsername());
        name.set(user.getName());
        phone.set(user.getPhone());
        address.set(user.getAddress());
        password.set(user.getPassword());
        genre.set(user.getGenre());
        avatarUrl.set(user.getAvatarUrl());
    }

    @SuppressLint("CheckResult")
    public void save() {
        if(isInfoFormValid()) {
            User newUserData = new User();
            newUserData.setAvatarUrl(avatarUrl.get());
            newUserData.setAddress(address.get());
            if(!user.getEmail().equals(email.get()))
                newUserData.setEmail(email.get());
            newUserData.setGenre(genre.get());
            newUserData.setName(name.get());
            newUserData.setPhone(phone.get());
            if(!user.getUsername().equals(username.get()))
                newUserData.setUsername(username.get());

            userModel.updateUser(newUserData, user.getObjectId())
                    .subscribe(next -> App.getInstance().setUser(next), error -> {
                        isLoading.set(false);
                        cleanupSubscriptions();
                        activityCallback.showDialogMessage(error.getMessage());
                    }, () -> {
                        isLoading.set(false);
                        cleanupSubscriptions();
                        activityCallback.finishCurrentActivity();
                    }, d -> {
                        isLoading.set(true);
                        disposable = d;
                    });
        }
    }

    public boolean isInfoFormValid() {
        if(TextUtils.isEmpty(name.get())) {
            nameFieldError.setErrorResource(R.string.required_field);
            return false;
        }

        if(TextUtils.isEmpty(email.get())) {
            emailFieldError.setErrorResource(R.string.required_field);
            return false;
        }

        if(!Patterns.EMAIL_ADDRESS.matcher(email.get()).matches()) {
            emailFieldError.setErrorResource(R.string.invalid_email_format);
            return false;
        }

        if(TextUtils.isEmpty(username.get())) {
            usernameFieldError.setErrorResource(R.string.required_field);
            return false;
        }

        return true;
    }

    public void onGenreSelected(AdapterView<?> parent, View view, int position, long id) {
        genre.set((String)parent.getSelectedItem());
    }

    public void changeAvatarImg() {
        avatarUrl.set(Constants.Other.ROBOHASH_API + new Random().nextInt(500) + Constants.Other.ROBOHASH_SET_2_PARAM);
    }

    @Override
    protected void onCleared() {
        cleanupSubscriptions();
        super.onCleared();
    }

    @Override
    public void cleanupSubscriptions() {
        if(disposable != null && !disposable.isDisposed())
            disposable.dispose();
    }
}

File: javaproject/app/src/main/java/com/marcosvbras/robomarket/flows/profile/viewmodel/EditProfileViewModel.kt
package com.marcosvbras.robomarket.flows.profile.viewmodel

import android.annotation.SuppressLint
import android.databinding.ObservableField
import android.text.TextUtils
import android.text.TextWatcher
import android.util.Patterns
import android.view.View
import android.widget.AdapterView
import com.marcosvbras.robomarket.R
import com.marcosvbras.robomarket.app.App
import com.marcosvbras.robomarket.business.domain.User
import com.marcosvbras.robomarket.business.model.UserModel
import com.marcosvbras.robomarket.interfaces.BaseActivityCallback
import com.marcosvbras.robomarket.utils.Constants
import com.marcosvbras.robomarket.utils.ErrorObservable
import com.marcosvbras.robomarket.utils.MaskWatcher
import com.marcosvbras.robomarket.viewmodels.BaseViewModel
import io.reactivex.disposables.Disposable
import java.util.*

class EditProfileViewModel(private val callback: BaseActivityCallback) : BaseViewModel() {

    private var disposable: Disposable? = null
    private var user: User? = null
    private val userModel: UserModel
    val phoneNumberWatcher = MaskWatcher.nineDigitPhone
    var email = ObservableField<String>()
    var username = ObservableField<String>()
    var name = ObservableField<String>()
    var phone = ObservableField<String>()
    var address = ObservableField<String>()
    var password = ObservableField<String>()
    var genre = ObservableField<String>()
    var avatarUrl = ObservableField<String>()
    var emailFieldError = ErrorObservable()

    var usernameFieldError = ErrorObservable()
    var nameFieldError = ErrorObservable()

    private val isInfoFormValid: Boolean
        get() {
            if (TextUtils.isEmpty(name.get())) {
                nameFieldError.errorResource = R.string.required_field
                return false
            }

            if (TextUtils.isEmpty(email.get())) {
                emailFieldError.errorResource = R.string.required_field
                return false
            }

            if (!Patterns.EMAIL_ADDRESS.matcher(email.get()!!).matches()) {
                emailFieldError.errorResource = R.string.invalid_email_format
                return false
            }

            if (TextUtils.isEmpty(username.get())) {
                usernameFieldError.errorResource = R.string.required_field
                return false
            }

            return true
        }

    init {
        setUserData()
        userModel = UserModel()
    }

    private fun setUserData() {
        val (objectId, name1, email1, username1, password1, emailVerified, sessionToken, avatarUrl1, phone1, genre1, address1) = App.getInstance().user
        user = User()
        user!!.address = address1
        user!!.username = username1
        user!!.phone = phone1
        user!!.name = name1
        user!!.genre = genre1
        user!!.email = email1
        user!!.avatarUrl = avatarUrl1
        user!!.emailVerified = emailVerified
        user!!.objectId = objectId
        user!!.password = password1
        user!!.sessionToken = sessionToken

        email.set(user!!.email)
        username.set(user!!.username)
        name.set(user!!.name)
        phone.set(user!!.phone)
        address.set(user!!.address)
        password.set(user!!.password)
        genre.set(user!!.genre)
        avatarUrl.set(user!!.avatarUrl)
    }

    @SuppressLint("CheckResult")
    fun save() {
        if (isInfoFormValid) {
            val newUserData = User()
            newUserData.avatarUrl = avatarUrl.get()
            newUserData.address = address.get()
            if (user!!.email != email.get())
                newUserData.email = email.get()
            newUserData.genre = genre.get()
            newUserData.name = name.get()
            newUserData.phone = phone.get()
            if (user!!.username != username.get())
                newUserData.username = username.get()

            userModel.updateUser(newUserData, user!!.objectId!!)!!
                    .subscribe({ next -> App.getInstance().user = next }, { error ->
                        isLoading.set(false)
                        cleanupSubscriptions()
                        callback.showDialogMessage(error.message!!)
                    }, {

                        cleanupSubscriptions()
                        callback.finishCurrentActivity()
                    }, { d ->
                        isLoading.set(true)
                        disposable = d
                    })
        }
    }

    fun onGenreSelected(parent: AdapterView<*>, view: View, position: Int, settedStr: String) {
        if(parent.getItemAtPosition(position).toString() != settedStr)
            genre.set(parent.selectedItem as String)
    }

    fun changeAvatarImg() {
        avatarUrl.set(Constants.Other.ROBOHASH_API + Random().nextInt(500) + Constants.Other.ROBOHASH_SET_2_PARAM)
    }

    override fun onCleared() {
        cleanupSubscriptions()
        super.onCleared()
    }

    override fun cleanupSubscriptions() {
        isLoading.set(false)
        disposable?.dispose()
    }
}